collect_app/src/main/java/org/odk/collect/android/application/initialization/SavepointsImporter.kt

```diff
@@ -0,0 +1,75 @@
+import org.odk.collect.forms.savepoints.SavepointsRepository
+import org.odk.collect.projects.ProjectsRepository
+import org.odk.collect.settings.keys.MetaKeys
+import org.odk.collect.upgrade.Upgrade
+import java.io.File
+
+class SavepointsImporter(
+    private val projectsRepository: ProjectsRepository,
+    private val instancesRepositoryProvider: InstancesRepositoryProvider,
+    private val formsRepositoryProvider: FormsRepositoryProvider,
+    private val savepointsRepositoryProvider: SavepointsRepositoryProvider,
+    private val storagePathProvider: StoragePathProvider
+) : Upgrade {
+    override fun key(): String {
+        return MetaKeys.OLD_SAVEPOINTS_IMPORTED
+    }
+
+    override fun run() {
+        projectsRepository.getAll().forEach { project ->
+            val instancesRepository = instancesRepositoryProvider.get(project.uuid)
+            val formsRepository = formsRepositoryProvider.get(project.uuid)
+            val savepointsRepository = savepointsRepositoryProvider.get(project.uuid)
+            val cacheDir = File(storagePathProvider.getOdkDirPath(StorageSubdirectory.CACHE, project.uuid))
+            val instancesDir = File(storagePathProvider.getOdkDirPath(StorageSubdirectory.INSTANCES, project.uuid))
+
+            importSavepointsThatBelongToSavedForms(instancesRepository, formsRepository, savepointsRepository, cacheDir)
+
+            importSavepointsThatBelongToBlankForms(formsRepository, savepointsRepository, cacheDir, instancesDir)
+        }
+    }
+
+    private fun importSavepointsThatBelongToSavedForms(instancesRepository: InstancesRepository, formsRepository: FormsRepository, savepointsRepository: SavepointsRepository, cacheDir: File) {
+        instancesRepository.all.forEach { instance ->
+            if (instance.deletedDate == null) {
+                val savepointFile = File(cacheDir, File(instance.instanceFilePath).name.plus(".save"))
+                if (savepointFile.exists() && savepointFile.lastModified() > instance.lastStatusChangeDate) {
+                    formsRepository.getAllByFormIdAndVersion(instance.formId, instance.formVersion).firstOrNull()?.let { form ->
+                        if (!form.isDeleted) {
+                            savepointsRepository.save(Savepoint(form.dbId, instance.dbId, savepointFile.absolutePath, instance.instanceFilePath))
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun importSavepointsThatBelongToBlankForms(formsRepository: FormsRepository, savepointsRepository: SavepointsRepository, cacheDir: File, instancesDir: File) {
+        formsRepository.all.sortedByDescending { form -> form.date }.forEach { form ->
+            if (!form.isDeleted) {
+                val formFileName = File(form.formFilePath).name.substringBeforeLast(".xml")
+
+                cacheDir.listFiles { file ->
+                    file.name.startsWith("${formFileName}_") && file.name.endsWith(".xml.save")
```

Comment:
It's easy to import savepoints that belong to saved forms because their names match `instanceFilePaths`. When it comes to blank forms it's a little bit cumbersome. We need to find savepoints with names starting with `formFilePaths` (we don't know the date of creation of a savepoint that might belong to our blank form which is a part of the savepoint file name so we find files based on their prefix and suffix).